### PR TITLE
gfx906/v0.10.1 MXFP4, MoE Tuning Fix and add GPT-OSS harmony flag for performance check

### PR DESCRIPTION
**Pull request by tomylin890**
## Summary
- Fix MXFP4 fallback on gfx906: keep scales in `uint8` on the right device, cast router logits, and dequant via `fused_experts`, so MXFP4 checkpoints run in FP16 emulation without Quark errors (requires `pip install amd-quark` and `apt-get install -y libxml2` in the current `nlzy/vllm-gfx906` Docker image).
- Make the GPT-OSS MoE router pick a dtype supported on ROCm and cast inputs/outputs to avoid BF16/Half mismatches.
- Add `VLLM_DISABLE_GPTOSS_HARMONY` so GPT-OSS reasoning wrappers can be toggled; the config no longer forces the parser when the flag is set, and both OpenAI front-ends skip Harmony in that case.
- Update `benchmark_moe.py` to honor `ROCR_VISIBLE_DEVICES`/`CUDA_VISIBLE_DEVICES`, pin Ray workers to `cuda:0`, and guard kernel launches with the correct device context.

## Testing
- `benchmarks/kernels/benchmark_moe.py`  
  - Tuned `Qwen3-30B-A3B-Thinking-2507-GPTQ-Int8` with `--enable-expert-parallel --tensor-parallel-size 4`; generated `E=32,N=768,device_name=Radeon_Instinct_MI50_32GB,dtype=int8_w8a16.json`.
- vLLM serve on MI50 (gfx906)  
  - MXFP4 checkpoints load and infer in FP16 emulation mode (after installing `amd-quark` and `libxml2`).  
  - `VLLM_DISABLE_GPTOSS_HARMONY=0/1` flips GPT-OSS responses between reasoning-wrapped and plain text (this adjustment is for some benchmarks script).
- Performance: at ~5120-token prompts, tuning improves prefill throughput from 1963 tok/s → 2326 tok/s and decode throughput from 39 tok/s → 47 tok/s (~20% uplift).

### CODE REVIEW

**Comment by Said-Akbar** [comment]
Hi @tomylin890 ,

Thank you for creating this PR to support gpt-oss. Can you please share if you used 2xMI50 32GB or 4x for your tests?
Also, when tuning, did you use batch_size=1 ? Thanks!

**Comment by Said-Akbar** [comment]
Can you please share your tuned output file?
This file: E=32,N=768,device_name=Radeon_Instinct_MI50_32GB,dtype=int8_w8a16.json

**Comment by tomylin890** [comment]
> Hi @tomylin890 ,
> 
> Thank you for creating this PR to support gpt-oss. Can you please share if you used 2xMI50 32GB or 4x for your tests? Also, when tuning, did you use batch_size=1 ? Thanks!

I tested gpt-oss-20B on 4xMI50, there is performance test results:
comment：gpt-oss-20b MXFP4 4xMI50 without json config (not generated)

| prompt length (tokens) | prefill time (ms) | prefill speed (tokens/s) | output length (tokens) | output time (ms) | output speed (tokens/s) |
|---------------------|-------------------|-------------------------|-------------------|---------------|---------------------|
| 512 | 464.00 | 1103.45 | 126 | 6681.00 | 18.86 |
| 5632 | 3761.00 | 1497.47 | 126 | 6819.00 | 18.48 |
| 10752 | 8165.00 | 1316.84 | 126 | 6952.00 | 18.12 |
| 15872 | 13404.00 | 1184.12 | 126 | 7094.00 | 17.76 |
| 20992 | 19290.00 | 1088.23 | 126 | 7241.00 | 17.40 |
| 26112 | 26013.00 | 1003.81 | 126 | 7339.00 | 17.17 |
| 31232 | 33711.00 | 926.46 | 126 | 7485.00 | 16.83 |
| 36352 | 42052.00 | 864.45 | 126 | 7628.00 | 16.52 |
| 41472 | 51667.00 | 802.68 | 126 | 7746.00 | 16.27 |

I think still have to do some optimization to speed up prefill and decode.

batch_size is defined in benchmark_moe.py:

`batch_sizes = [
            1,
            2,
            4,
            8,
            16,
            24,
            32,
            48,
            64,
            96,
            128,
            256,
            512,
            1024,
            1536,
            2048,
            3072,
            4096,
        ]`

So, if you generate json config file by this script, it has batch_size=1.

**Comment by tomylin890** [comment]
> Can you please share your tuned output file? This file: E=32,N=768,device_name=Radeon_Instinct_MI50_32GB,dtype=int8_w8a16.json

Here is json config I generated on my platform (5700X + 4xMI50 under PEX88096 PCIe Switch, VBIOS 113-D1631700-111, PCIe P2P is enabled)

Model: https://huggingface.co/QuantTrio/Qwen3-30B-A3B-Thinking-2507-GPTQ-Int8

```
{
    "1": {
        "BLOCK_SIZE_M": 16,
        "BLOCK_SIZE_N": 16,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 1,
        "num_warps": 2,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "2": {
        "BLOCK_SIZE_M": 16,
        "BLOCK_SIZE_N": 16,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 1,
        "num_warps": 2,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "4": {
        "BLOCK_SIZE_M": 16,
        "BLOCK_SIZE_N": 16,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 1,
        "num_warps": 2,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "8": {
        "BLOCK_SIZE_M": 16,
        "BLOCK_SIZE_N": 64,
        "BLOCK_SIZE_K": 64,
        "GROUP_SIZE_M": 1,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "16": {
        "BLOCK_SIZE_M": 16,
        "BLOCK_SIZE_N": 16,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 1,
        "num_warps": 2,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "24": {
        "BLOCK_SIZE_M": 16,
        "BLOCK_SIZE_N": 16,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 8,
        "num_warps": 2,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "32": {
        "BLOCK_SIZE_M": 16,
        "BLOCK_SIZE_N": 32,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 8,
        "num_warps": 4,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "48": {
        "BLOCK_SIZE_M": 16,
        "BLOCK_SIZE_N": 128,
        "BLOCK_SIZE_K": 64,
        "GROUP_SIZE_M": 1,
        "num_warps": 2,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "64": {
        "BLOCK_SIZE_M": 16,
        "BLOCK_SIZE_N": 64,
        "BLOCK_SIZE_K": 64,
        "GROUP_SIZE_M": 1,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "96": {
        "BLOCK_SIZE_M": 32,
        "BLOCK_SIZE_N": 64,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 1,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "128": {
        "BLOCK_SIZE_M": 32,
        "BLOCK_SIZE_N": 64,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 1,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "256": {
        "BLOCK_SIZE_M": 32,
        "BLOCK_SIZE_N": 64,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 1,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "512": {
        "BLOCK_SIZE_M": 32,
        "BLOCK_SIZE_N": 64,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 8,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "1024": {
        "BLOCK_SIZE_M": 32,
        "BLOCK_SIZE_N": 64,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 8,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "1536": {
        "BLOCK_SIZE_M": 64,
        "BLOCK_SIZE_N": 32,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 8,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "2048": {
        "BLOCK_SIZE_M": 128,
        "BLOCK_SIZE_N": 16,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 4,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "3072": {
        "BLOCK_SIZE_M": 128,
        "BLOCK_SIZE_N": 16,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 1,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    },
    "4096": {
        "BLOCK_SIZE_M": 128,
        "BLOCK_SIZE_N": 16,
        "BLOCK_SIZE_K": 32,
        "GROUP_SIZE_M": 4,
        "num_warps": 1,
        "num_stages": 2,
        "waves_per_eu": 0
    }
}
```

Some News: I also make Qwen3Next work on nlzy's repo. But still optimizing, there is some performance test:

comment: Qwen3-Next-80B-A3B-Thinking-AWQ-4bit 4xMI50 without json config (not generated)

| prompt length (tokens) | prefill time (ms) | prefill speed (tokens/s) | output length (tokens) | output time (ms) | output speed (tokens/s) |
|---------------------|-------------------|-------------------------|-------------------|---------------|---------------------|
| 512 | 542.00 | 944.65 | 128 | 5501.00 | 23.27 |
| 5632 | 4460.00 | 1262.78 | 128 | 6125.00 | 20.90 |
| 10752 | 10364.00 | 1037.44 | 128 | 6765.00 | 18.92 |
| 15872 | 17625.00 | 900.54 | 128 | 7414.00 | 17.26 |
| 20992 | 27014.00 | 777.08 | 128 | 8016.00 | 15.97 |
| 26112 | 37575.00 | 694.93 | 128 | 9243.00 | 13.85 |
| 31232 | 50489.00 | 618.59 | 128 | 9898.00 | 12.93 |
| 36352 | 64395.00 | 564.52 | 128 | 10501.00 | 12.19 |
| 41472 | 80696.00 | 513.93 | 128 | 11120.00 | 11.51 |

 I think it also needs to be optimized too.., It is a big work because I choose backport instead of update to upstream vllm v0.10.2.

**Comment by Said-Akbar** [comment]
Great! Thanks! Looking forward to those updates.

**Comment by iEddie-cmd** [comment]
> comment: Qwen3-Next-80B-A3B-Thinking-AWQ-4bit 4xMI50 without json config (not generated)
> 
> prompt length (tokens)	prefill time (ms)	prefill speed (tokens/s)	output length (tokens)	output time (ms)	output speed (tokens/s)
> 512	542.00	944.65	128	5501.00	23.27
> 5632	4460.00	1262.78	128	6125.00	20.90
> 10752	10364.00	1037.44	128	6765.00	18.92
> 15872	17625.00	900.54	128	7414.00	17.26
> 20992	27014.00	777.08	128	8016.00	15.97
> 26112	37575.00	694.93	128	9243.00	13.85
> 31232	50489.00	618.59	128	9898.00	12.93
> 36352	64395.00	564.52	128	10501.00	12.19
> 41472	80696.00	513.93	128	11120.00	11.51
> I think it also needs to be optimized too.., It is a big work because I choose backport instead of update to upstream vllm v0.10.2.

How did you made Qwen3-Next run? I get a Value error, The checkpoint you are trying to load has model type `qwen3_next` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date. error.

**Comment by tomylin890** [comment]
> > comment: Qwen3-Next-80B-A3B-Thinking-AWQ-4bit 4xMI50 without json config (not generated)
> > prompt length (tokens)	prefill time (ms)	prefill speed (tokens/s)	output length (tokens)	output time (ms)	output speed (tokens/s)
> > 512	542.00	944.65	128	5501.00	23.27
> > 5632	4460.00	1262.78	128	6125.00	20.90
> > 10752	10364.00	1037.44	128	6765.00	18.92
> > 15872	17625.00	900.54	128	7414.00	17.26
> > 20992	27014.00	777.08	128	8016.00	15.97
> > 26112	37575.00	694.93	128	9243.00	13.85
> > 31232	50489.00	618.59	128	9898.00	12.93
> > 36352	64395.00	564.52	128	10501.00	12.19
> > 41472	80696.00	513.93	128	11120.00	11.51
> > I think it also needs to be optimized too.., It is a big work because I choose backport instead of update to upstream vllm v0.10.2.
> 
> How did you made Qwen3-Next run? I get a Value error, The checkpoint you are trying to load has model type `qwen3_next` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date. error.

I am backporting Qwen3Next in the current v0.10.1 vllm branch of nlzy. This is a very large workload (Hybrid Attention/mamba, and there are many dependency issues to deal with. This is much more complicated than writing an FP16 simulation path for gpt-oss. 

I finally encountered the problem of prefill. The first Q&A was fine, but the second Q&A was garbled, and this problem spread to other categories of models (MXFP4/GPTQ/AWQ), so I gave up the working source tree. Maybe I will establish a v0.10.2 branch in the near future. 

I will merge the upstream branch and keep all the optimizations of MI50. When I confirm that the existing model is working well, I will develop the Qwne3Next part again, which should be much simpler. 

But I am now facing the problem of environmental dependency because vllm 0.10.2 requires the following packages:
`
torch==2.8.0
torchvision==0.23.0
torchaudio==2.8.0
triton==3.3.0
`

So I had to upgrade my environment to Torch 2.8 for development, but I ran

`pip3 install 'torch==2.8' torchvision torchaudio --index-url https://download.pytorch.org/whl/rocm6.3`

In addition to torch, torchvision, and torchaudio, pip also updated pytorch-triton-rocm 3.4.0. I'm not sure if this is a pytorch-triton-rocm issue, as I encountered an "unsupported target" error when running models with the existing nlzy v0.10.1 branch. 

The issue with 'gfx906' (triton is using nlzy version 3.3.0, and even if I manually downgrade to 3.3.0, I still encounter this issue, and the downgrade triggers a pip dependency warning) has not yet been resolved, so I can't start development yet (although it may be a basic and simple issue).

If anyone can help with the environment dependencies, I hope I can start merging upstream and running model development as soon as possible.

p.s. I am running ROCm 6.3.3.

**Comment by nlzy** [comment]
> pip also updated pytorch-triton-rocm 3.4.0. I'm not sure if this is a pytorch-triton-rocm issue, as I encountered an "unsupported target" error when running models with the existing nlzy v0.10.1 branch.

It should be pip downloaded offical triton v3.4.0 from pypi for you, so got a "unsupported target" error.

I just updated triton to 3.4.0. Give it a try: https://github.com/nlzy/triton-gfx906/tree/gfx906/v3.4.x

**Comment by nlzy** [comment]
Great job!

Merged directly in https://github.com/nlzy/vllm-gfx906/commit/6dce86591ed66b76b5d56a6f10d562559eb61622
